packages/fuels-contract/src/contract.rs

```diff
@@ -487,35 +257,86 @@ impl Contract {
-    pub fuel_client: FuelClient,
+pub struct ContractCall {
+    pub tx_parameters: TxParameters,
+    pub wallet: LocalWallet,
+    pub contract_id: ContractId,
     pub encoded_args: Vec<u8>,
     pub encoded_selector: Selector,
-    pub contract_id: ContractId,
-    pub tx_parameters: TxParameters,
     pub call_parameters: CallParameters,
     pub maturity: u64,
-    pub datatype: PhantomData<D>,
-    pub output_params: Vec<ParamType>,
     pub compute_calldata_offset: bool,
-    pub wallet: LocalWallet,
     pub variable_outputs: Option<Vec<Output>>,
-    external_contracts: Option<Vec<ContractId>>,
+    pub external_contracts: Option<Vec<ContractId>>,
+    pub output_params: Vec<ParamType>,
+}
+
+impl ContractCall {
+    /// Based on the returned Contract's output_params and the receipts returned from the call,
+    /// decode the values and return them.
+    pub fn get_decoded_output(
+        &self,
+        mut receipts: Vec<Receipt>,
+    ) -> Result<(Vec<Token>, Vec<Receipt>), Error> {
+        // Multiple returns are handled as one `Tuple` (which has its own `ParamType`), so getting
+        // more than one output param is an error.
+        if self.output_params.len() != 1 {
+            return Err(Error::InvalidType(format!(
+                "Received too many output params (expected 1 got {})",
+                self.output_params.len()
+            )));
+        }
+        let output_param = self.output_params[0].clone();
+
+        let (encoded_value, index) = match output_param.get_return_location() {
+            ReturnLocation::ReturnData => {
+                match receipts.iter().find(|&receipt| receipt.data().is_some()) {
+                    Some(r) => {
+                        let index = receipts.iter().position(|elt| elt == r).unwrap();
+                        (r.data().unwrap().to_vec(), Some(index))
+                    }
+                    None => (vec![], None),
+                }
+            }
+            ReturnLocation::Return => {
+                match receipts.iter().find(|&receipt| receipt.val().is_some()) {
+                    Some(r) => {
+                        let index = receipts.iter().position(|elt| elt == r).unwrap();
+                        (r.val().unwrap().to_be_bytes().to_vec(), Some(index))
+                    }
+                    None => (vec![], None),
+                }
+            }
+        };
+        if let Some(i) = index {
+            receipts.remove(i);
+        }
+        let mut decoder = ABIDecoder::new();
+        let decoded_value = decoder.decode(&self.output_params, &encoded_value)?;
+        Ok((decoded_value, receipts))
+    }
 }
 
-impl<D> ContractCall<D>
+#[derive(Debug)]
+#[must_use = "contract calls do nothing unless you `call` them"]
+/// Helper for managing a transaction before submitting it to a node
+pub struct ContractCallHandler<D> {
+    pub contract_call: ContractCall,
+    pub fuel_client: FuelClient,
+    pub datatype: PhantomData<D>,
```

Comment:
what is this `PhantomData` (I know it's not from your changes but it should have been documented)

Reply:
Rust complains if `D` is not used in a data field so it is just a placeholder to make the compiler happy. I think docs are unnecessary since the Rust docs cover the use of `PhantomData`